community/codegen/src/main/java/org/neo4j/codegen/bytecode/ByteCodeExpressionVisitor.java

```diff
@@ -325,7 +325,7 @@ public void notEqual( Expression lhs, Expression rhs )
 
     private void equal( Expression lhs, Expression rhs, boolean equal )
     {
-        assertSameType( lhs, rhs, "compare" );
+//        assertSameType( lhs, rhs, "compare" ); // TODO: Is this assert unnecessary / too strong?
```

Comment:
In byte code you can only compare exact types so this assertion is valid I think

Reply:
I think this is true for some primitives since they have different compare instruction, but hardly for reference types? If I enable this assertion I cannot compare an `AnyValue` to a `Value`, but without it that yields correctly working bytecode.